File: lib/kemperstomp/src/controller/PeriodCounter.py
from ..misc.Tools import Tools

# Periodic update helper    
class PeriodCounter:
    def __init__(self, interval_millis):
        self._interval_millis = int(interval_millis)

        self._last_reset = 0

    # Returns if the period has been exceeded. If yes, it lso resets
    # the period to the current time.
    @property
    def exceeded(self):
        current_time = Tools.get_current_millis()
        if self._last_reset + self._interval_millis < current_time:
            self._last_reset = current_time
            return True
        return False
            

File: web/htdocs/js/misc/PeriodCounter.js
class PeriodCounter {

    interval = null;

    #lastReset = 0;
    #overrideTimeCallback = null;

    constructor(intervalMillis, overrideTimeCallback = null) {
        this.interval = intervalMillis;
        this.#overrideTimeCallback = overrideTimeCallback;

        this.reset();
    }
    
    /**
     * Resets the period counter to the current time
     */
    reset() {
        this.#lastReset = this.#getCurrentTimestamp();
    }

    /**
     * Returns the amount of milliseconds passed since the last reset
     */
    passed() {
        return this.#getCurrentTimestamp() - this.#lastReset;
    }

    /**
     * Returns if the period has been exceeded. If yes, it lso resets 
     * the period to the current time.
     */
    exceeded() {
        const currentTime = this.#getCurrentTimestamp();
        if (this.#lastReset + this.interval < currentTime) {
            this.#lastReset = currentTime;
            return true;
        }
        return false;
    }

    /**
     * Returns current unix timestamp
     */
    #getCurrentTimestamp() {
        if (this.#overrideTimeCallback) {
            return this.#overrideTimeCallback();
        }
        return Date.now();
    }
}